### ISSUE TITLE
The lack of varity in anti-missile point defence systems

### ISSUE THREAD

**Issue by gunqqer** [endless-sky/endless-sky#4378]
Currently all our anti missile systems are just stat swaps. Some have faster reload but are worse at shooting things down, and vice verse, etc. But that's it, they all shoot down missiles with a hitscan beam. There's as I said in the title, no varity
My favorite solution: Make it so any weapon type can shoot down missiles. Now I'm not saying just let any weapon that is already in the game shoot down missiles, but make it so there's a tag that means any created weapon can shoot down missiles. So with this you could have anti-missile missiles, flak weaponry, etc. This would require someone figuring out how targeting works. Will weaponry with tracking re-target if its target is taken out? Will we allow these weapons to do damage to ships (see: #2319 )? And probably many more things I haven't though of

**Comment by solardawning** [comment]
Missile strength would have to be re-imagined as a Hull/HP stat.

Weapons would need targeting rules and priority order, such as: 
"Can Target: Ship, Fightercraft, Missile"
"Target Priority Order: "Missile, Fightercraft, Ship"

Individual weapons, as they could have different valid targets and priority orders, would need their own target logic.

I've seen these sorts of systems work very well, but it's certainly a can of worms.

**Comment by Zitchas** [comment]
I've always thought that the hitscan beam style thing was more of a simple work around rather than something that should be permanent. Kind of like how all guns shouldn't be beam weapons, although the beams are easier to deal with.

Anyway, I'm definitely in favor of this, along with the targeting limiters.

I would also like to point out that this is how they worked in EVN. Missiles and the like were projectiles that could be actively targeted by weapons (usually automatic things similar to our point defense stuff), and the point defense stuff was typically more like traditional AAA batteries rather than lasers. Well, core human space was various kinds of AAA batteries (which often doubled as anti-fighter turrets), with other areas of space having other things.

Anyway, this is a good opportunity to make anti-missile defense more varied, potentially more useful, as well as change up the dynamic for both fighters and missiles a bit; which are both things that need work.

**Comment by tehhowch** [comment]
In the past, I looked briefly into what it requires to do something like chaff or flares (essentially projectile retargeting), and it can get nasty for performance if not implemented well.

Having any projectile able to deal damage to homing projectiles introduces a large balance change (Why add a dedicated A/M? Just add more lasers that can shoot ships, too!) that makes do-everything builds more possible. That's not to say it can't be done, just that it needs to be considered carefully.

**Comment by gunqqer** [comment]
Yeah 100% most weapons that are for fighting ships shouldn't be able to do AM, or if they do, very very very badly. It's more of, don't restrict the ability (especially so plugin creators can do what they want, as we are not worried about how plugin creators balance stuff). Chaffs/flares aren't really what I was thinking, but one way of fixing the performance hit is to instead of having the missiles retarget to the flares to just have them keep going in the same direction or have them go haywire.

**Comment by Zitchas** [comment]
In regards to the performance impact, I'm kind of curious about that. I don't understand the code, but I know that the original EVN did do exactly this: Calculate all projectiles as individual objects like ships that were targeted by auto-firing turrets, certain kinds (infrared guidance) were subject to being lured off track by flares; any weapon could theoretically be given the flag that made their shots damage missiles, etc. And that ran quite smoothly on generic iMacs back in 2002. The only real limiter that I can tell is that in EVN the player was limited to 6 escorts, and I believe a system couldn't have more than 36 actual ships in it at a time. Obviously, ES doesn't have those restrictions, so I can definitely see it piling up for larger engagements. While I don't understand the code, I'm curious as to whether or not EV had more efficient code or something to run nicely with less available horsepower.

In regards to having any projectile/beam being able to damage a projectile, it wasn't actually that. Any projectile/beam *could* have a flag set (or in other words, an attribute) that meant it would damage a missile. In practice, only the weapons described in-game as Anti-Air-Artillery or Anti-Missile weapons had the attribute set. In theory, one could make a plug-in that gave that attribute to everything, but that had its own problems. Namely the auto-fire.

While generally speaking EVN had no automation for firing weapons, the game *did* have built-in automation for everything with the AAA attribute, so in practice, the game felt very much like it does in ES. You fly around, and when a missile gets close, the AAA starts lighting it up, hopefully resulting in a little explosion. The big difference was that AAA would also target fighters, not just missiles. So loading a ship with lots of AAA would make a really good anti-fighter & anti-missile platform, but would end up being very weak against pretty much everything else. 

In human space, the predominant AAA was the [Quad Light Blaster Turret](https://evn.fandom.com/wiki/Quad_Light_Blaster_Turret), which was, as you might expect, a very fast firing phalanx style weapon that tracked very quickly and fired a lot of shots, but generally did pretty low damage. It was reasonably effective against both fighters and missiles in human space, but fairly week against the rest of the galaxy.

The flip side of all this, however, was that (according to my fuzzy memory) fighters tended to be smaller compared to other ships than they are in ES, and both main turrets and ships in general both turned a bit slower. (and there were fewer ways to speed up ship turning) So dealing with enemy fighters manually usually meant trying to maneuver around so that you could line up your guns with their path as they did an attack run. Fighters died pretty quickly to that, if one could manage it. Otherwise, taking out fighters with turrets kind of  felt like trying to shoot down fighters with a battleship main guns: Sure, if a shot connects that fighter is *toast*, but it takes a fair bit of luck or skill to actually make that connection. 

As a result, having dedicated AAA turrets was probably just as important to have as it is in ES, perhaps more so since it was quite important for dealing with fighters, but there were few enough AAA weapons that dealt any significant damage that a ship with predominantly AAA wasn't particularly feasible as a generalist platform. (the Manticore and a few others did have AAA versions, and they were generally only really effective as part of a fleet. On their own the suffered compared to their normal versions.) It did, however, feel much more reaslistic; and was very interesting to watch. It was a lot of fun watching enemy fighters start an attack run and see the cloud of light blaster bolts firing at them. Looked very reminiscent of fighter combat scenes in various Star Wars movies.

**Comment by Tadrix** [comment]
> It's more of, don't restrict the ability (especially so plugin creators can do what they want, as we are not worried about how plugin creators balance stuff).

Except in this case you aren't asking to lift restrictions; you are asking to put a feature in the game that doesn't need it (vanilla, at least). All such cases need to be considered very carefully, as, once put in, it has to be supported in perpetuity (through all code refactoring instances and what not, which means that improving the code becomes increasingly difficult the more such features are there), for no apparent benefit for the game itself.

**Comment by gunqqer** [comment]
`For no apparent benefit`
No there is benifit. Adding more fuinctions to the game will add variety and new ways of having anti-missile work. Currently everything is just stat swaps. Which means there is very limited variety. I want that to change.
And by "restriction" I was speaking more abstractly that you think. It isn't like we specifically said you shouldn't be able to do this, but it is restricted as in you can't do it.

**Comment by jafdy** [comment]
Creating an anti-missile projectile weapon has already been noed (the calculations are too hard apparently). Allowing all other weapons to just so happen to hit missiles would be tricky as then you would need to define an anti missile value for them.

**Comment by gunqqer** [comment]
If it was such a long time ago, we should revist it. Also, I could be wrong, but colisions are pretty easy/low CPU usage. Like yeah if you have a system full of missiles flying, and other missiles flying at those missiles it could get dicey but I fail to see how it would be difficult. We already have missiles that track ships, how different can it be to tracka missile?

**Comment by ChamEV** [comment]
It's because the # of collision checks grows quadratically, though there are algorithms that can alleviate this, they add complexity.

**Comment by Tadrix** [comment]
> If it was such a long time ago, we should revist it.

Indeed, if something fundamentally changed. But nothing has changed in this regard. Past considerations would still be valid here.

**Comment by gunqqer** [comment]
> > If it was such a long time ago, we should revist it.
> 
> Indeed, if something fundamentally changed. But nothing has changed in this regard. Past considerations would still be valid here.

Given it was such "in the deep dark past" and "only a few of us can remembger" (jafdy) it seems like a lot will have changed with something that long ago. 



> It's because the # of collision checks grows quadratically, though there are algorithms that can alleviate this, they add complexity.

I know a lot of people like how very low computer power this game is, but we shouldn't limit outselfs like this. antimissiles only have a certain range, which for most other than missile is small, so it shouldn't massivly increase the need for CPU power.

**Comment by jafdy** [comment]
The only thing that has changed since then is the nature of CPU tech (mainly AMD resurgent). Because of the increase in collision calculations that would be needed we might need to multi-thread the game. It is also worth noting that most fights happen within that close range so ships could be spewing projectiles everywhere, just think of how many shots something like a stock protector fires!

**Comment by gunqqer** [comment]
> The only thing that has changed since then is the nature of CPU tech (mainly AMD resurgent). Because of the increase in collision calculations that would be needed we might need to multi-thread the game. It is also worth noting that most fights happen within that close range so ships could be spewing projectiles everywhere, just think of how many shots something like a stock protector fires!

"Now I'm not saying just let any weapon that is already in the game shoot down missiles, but make it so there's a tag that means any created weapon can shoot down missiles"

You misunderstand me, I don't want every weapon to have anti-missile, but have more possibilities for anti-missile some of which that can also do ship to ship

**Comment by Tadrix** [comment]
> I know a lot of people like how very low computer power this game is, but we shouldn't limit outselfs like this.

*cough* The Battle of Mesuket *cough*

Since this is GitHub: There are moments when this game already struggles on low-end systems. If you add operations needed in exponential amounts, then we might have a problem.

**Comment by SirStigII** [comment]
Isn't part of why mesuket struggles is exactly because it's not using multi threads?

Wouldn't it be easier to add an extra value to AM/fighters that would allow a percentage of damage based on AM strength?

**Comment by Zitchas** [comment]
I'm still curious as to how a game from 17 years ago managed to handle auto-firing AAA turrets that actually targeted and shot at fighters and missiles (and had to abide by all the normal considerations of dealing with things dodging, asteroids, etc like any other weapon) without lagging out anything. In EVN, insta-hit AAA fire (aka beams) was limited to top-tier AAA weaponry - most AAA weaponry was of the "spam hundreds of shots really fast and hope enough connect to kill the fighter/missile" type.  (aka light blasters and chainguns). And they actually were lots of shots, one could watch them, and dodge them if one was in a fighter. 

Might there not be some lesson for us, or something we can emulate from how they handled it? Even bottom-end computers these days should significantly exceed the average Performa or iMac from 17 years ago.


That aside, I also like the idea of having "antimissile" be a weapon attribute that specifies that said weapon will damage a missile if it should happen to come in contact with it. That doesn't need any extra tracking or auto-firing capabilities. The main difference is that missiles will need to have whatever calculations are needed to keep track of potential hits.

**Comment by Tadrix** [comment]
Might be the other way round. EVN & Co. have restrictions this game doesn't have. The amount of ships (escorts, in-system, etc., the scale is fixed), including derelicts, calculating asteroids, [which the EV series cheated with](https://github.com/endless-sky/endless-sky/wiki/TiledAsteroids), etc.

**Comment by Zitchas** [comment]
That's true, but it is good to list what the differences is.

For those who aren't aware, the EV series had a hard limit on number of ships in the system. 36, I think. (I'm not 100% on the exact number, but it was something close to that.)

And the game didn't render asteroids much passed the edge of the screen.

**Comment by Tadrix** [comment]
To be honest, I haven't played it, and information on the Internet is scarce (or perhaps I'm bad at finding it). I'm sure MZ is intimately familiar with its strengths and shortcomings and did the best job he could.

What little I've managed to find at a quick glance is this:

> Therefore, there are many limitations to plug-ins, most notably the lack of any sort of scripting, inability to add new features (such as A.I. types, weapon behaviors, etc.), limits on plug-in size (usually worked around by splitting a plug into multiple files), the requirement to use specific graphic and sound types (particularly antiquated System 7 format for sound effects), and seemingly arbitrary maximum and minimum resource values, all of which made perfect sense ten years ago.

[Source](https://evn.fandom.com/wiki/Plug-in#Developing_Plug-ins)

Another thing is that [EVN had its hiccups, even with its restrictions](http://www.ambrosiasw.com/support/faqs/products/ev-nova/Why-does-the-game-slow-down-when-there-are-many-ships-in-a-system).

**Comment by Blaank** [comment]
Having each race or faction favor a specific style of AMS would add some nice flavor

**Comment by Pointedstick** [comment]
It seems like there are two different conversations going on here:
1. Add support for anti-missile weapons that fire a stream of projectiles or anti-missile missiles
2. Hybrid defense turrets that can shoot at both missiles and fighters

It might make sense to separate these into their own issues for clarity.
